### PR TITLE
Fixed #2584 better error handling with Hash function

### LINKED ISSUE #2584: better error handling with Hash() (hash.Hash256, error)
**What would you like to be added**:
hash of envelope/sealed envelope only returns hash.Hash256, hiding any error that might occur during the process. as a result, panic is used (instead of an all-zero hash value) to indicate error

return (hash.Hash256, error) could mitigate this issue and remove unnecessary panic

**Why is this needed**:
better code

### PR DESCRIPTION
**Pull request by ashishsnigam**
Fixed #2584

1. Updated function in file /action/sealedenvelope.go to return error also. New signature will be func (sealed *SealedEnvelope) Hash() (hash.Hash256, error)
2. Updated calling of this function as per new signature across entire source code
3. Review comments incorporated given on PR Fixed #2584 - better error handling with Hash() (hash.Hash256, error) #2732
4. New PR due to cla error

### CODE REVIEW

**Comment by ashishsnigam** [comment]
@dustinxie @CoderZhi Please review this PR. All review comments are incorporated.

**Review by dustinxie — APPROVED** [review]

**Comment by dustinxie** [comment]
pls fix test failure in `integrity_test.go`

**Comment by ashishsnigam** [comment]
Fixed test. Sorry I did not notice your comments on sealedenvelop.go file. @dustinxie

File: actpool/actpool.go [hunk 1]
```diff
@@ -447,7 +454,10 @@ func (ap *actPool) removeConfirmedActs() {
 
 func (ap *actPool) removeInvalidActs(acts []action.SealedEnvelope) {
 	for _, act := range acts {
-		hash := act.Hash()
+		hash, err := act.Hash()
+		if err != nil {
+			log.L().Fatal("Failed to get hash", zap.Error(err))
```
[the remark this answers is not in the document]
Comment (ashishsnigam) [reply actpool/actpool.go:459]: skipping it as no way to return error from this function

File: actpool/actpool.go [hunk 2]
```diff
@@ -460,11 +470,15 @@ func (ap *actPool) removeInvalidActs(acts []action.SealedEnvelope) {
 // deleteAccountDestinationActions just for destination map
 func (ap *actPool) deleteAccountDestinationActions(acts ...action.SealedEnvelope) {
 	for _, act := range acts {
+		hash, err := act.Hash()
+		if err != nil {
+			log.L().Fatal("Failed to get hash", zap.Error(err))
```
[the remark this answers is not in the document]
Comment (ashishsnigam) [reply actpool/actpool.go:475]: skipped

File: api/api.go [hunk 3]
```diff
@@ -1411,7 +1420,10 @@ func (api *Server) actionsInBlock(blk *block.Block, start, count uint64) []*iote
 	}
 	for i := start; i < lastAction; i++ {
 		selp := blk.Actions[i]
-		actHash := selp.Hash()
+		actHash, err := selp.Hash()
+		if err != nil {
+			log.L().Fatal("Failed to get hash", zap.Error(err))
```
[the remark this answers is not in the document]
Comment (ashishsnigam) [reply api/api.go:1425]: skipped

File: api/api.go [hunk 4]
```diff
@@ -1436,7 +1448,10 @@ func (api *Server) reverseActionsInBlock(blk *block.Block, reverseStart, count u
 	for i := reverseStart; i < uint64(len(blk.Actions)) && i < reverseStart+count; i++ {
 		ri := uint64(len(blk.Actions)) - 1 - i
 		selp := blk.Actions[ri]
-		actHash := selp.Hash()
+		actHash, err := selp.Hash()
+		if err != nil {
+			log.L().Fatal("Failed to get hash", zap.Error(err))
```
[the remark this answers is not in the document]
Comment (ashishsnigam) [reply api/api.go:1453]: skipped

File: blockchain/block/block.go [hunk 5]
```diff
@@ -158,7 +159,10 @@ func (b *Block) TransactionLog() *BlkTransactionLog {
 func (b *Block) ActionHashs() []string {
 	actHash := make([]string, len(b.Actions))
 	for i := range b.Actions {
-		h := b.Actions[i].Hash()
+		h, err := b.Actions[i].Hash()
+		if err != nil {
+			log.L().Fatal("Failed to get hash", zap.Error(err))
```
[the remark this answers is not in the document]
Comment (ashishsnigam) [reply blockchain/block/block.go:164]: skipped

File: blockchain/block/utils.go [hunk 6]
```diff
@@ -20,7 +20,11 @@ import (
 func calculateTxRoot(acts []action.SealedEnvelope) hash.Hash256 {
 	h := make([]hash.Hash256, 0, len(acts))
 	for _, act := range acts {
-		h = append(h, act.Hash())
+		actHash, err := act.Hash()
+		if err != nil {
+			return hash.ZeroHash256
```
[the remark this answers is not in the document]
Comment (ashishsnigam) [reply blockchain/block/utils.go:25]: function is returning hash.Hash256 and not error, that's why hash.ZeroHash256. Not sure how to return error here. So no change in next commit.

**Comment by ashishsnigam** [comment]
@CoderZhi I have skipped in all scenarios as there was no way to return error from those functions. Please review it again.

File: blockchain/block/utils.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (dustinxie) [reply blockchain/block/utils.go:28]: I suggest:
1. we merge this PR for now
2. I've created a follow-up task: https://github.com/iotexproject/iotex-core/issues/2739 for @ashishsnigam to continue work on

[the remark this answers is not in the document]
Comment (ashishedjx) [reply blockchain/block/utils.go:28]: Sure. I can pick this followup task. @dustinxie

**Review by CoderZhi — APPROVED** [review]

**Comment by dustinxie** [comment]
@ashishsnigam  would you like to continue work on #2739?

**Comment by ashishsnigam** [comment]
> @ashishsnigam would you like to continue work on #2739?

Sure @dustinxie .

**Pull request merged**